source/common/upstream/cluster_manager_impl.cc

```diff
@@ -473,18 +473,52 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
     cluster_warming_cb(cluster_name, ClusterWarmingState::Starting);
     cluster_entry->cluster_->initialize([this, cluster_name, cluster_warming_cb] {
       auto warming_it = warming_clusters_.find(cluster_name);
-      auto& cluster_entry = *warming_it->second;
+      auto& warming_cluster_entry = *warming_it->second;
 
       // If the cluster is being updated, we need to cancel any pending merged updates.
       // Otherwise, applyUpdates() will fire with a dangling cluster reference.
       updates_map_.erase(cluster_name);
 
+      // If management server sends a EDS response, for any other cluster grpc_mux_impl calls
```

Comment:
@ramaraochavali @snowp @htuch you all are clearly fine with this, so I'm not understanding something, but why does an EDS response for any other cluster cause an empty update for this cluster? Isn't that a bug?

Reply:
OK I think I understand, we are looking at this text in the XDS protocol docs: "When a requested resource is missing in a RDS or EDS update, Envoy will retain the last known value for this resource." Right?

I guess this seems OK, though I do wonder if this clause is adding extra complexity that we don't really need, especially with incremental coming. Meaning, it seems like your management server could not accept new connections until it's ready to serve them and we can avoid all of this extra code, potential for bugs, behavior differences, etc. That would be my preference.

If we do decide to keep this, can we spell out the clause we are relying on here in the docs and link to them?

Reply:
@htuch knows better but IIRC it is required for initial loading otherwise Envoy was struck at initialization - I tried to change that behaviour at some point via https://github.com/envoyproxy/envoy/pull/4276 but got a regression https://github.com/envoyproxy/envoy/issues/4485 and reverted it via https://github.com/envoyproxy/envoy/pull/4490

Reply:
> we are looking at this text in the XDS protocol docs: "When a requested resource is missing in a RDS or EDS update, Envoy will retain the last known value for this resource." Right?

Yes.

> If we do decide to keep this, can we spell out the clause we are relying on here in the docs and link to them?

I can add the clause in the doc, but we need this fix on Envoy side - it should not clear hosts when it gets EDS response for some other cluster. For EDS/RDS we are not mandating to send all responses every time (Which is the right thing to do). Things may change with incremental but we can deal with it when it comes.

Reply:
@ramaraochavali we just discussed this issue briefly on the community call. I think we might need to have a meeting on this. Given that chat, I'm still pretty uncomfortable with this change for 2 reasons:
1) It's not clear to me that's correct to copy the hosts from an old cluster to a new cluster.
2) It adds a bunch of complexity that might lead to hard to understand edge cases.
3) It really seems to me like you could fix this in your management server: don't accept connections to a management server (do not report healthy to your LB) until the management server can fully serve connections.

I know @htuch as some thoughts here so will let him weigh in and then we can maybe do a meeting if needed?

Reply:
@mattklein123 Sure. Would be happy to chat if needed on this. 

Some clarifications.

> It really seems to me like you could fix this in your management server

IMO, that does not solve the problem completely (btw, we are on our way to do that already) since most of the discovery backends like Zookeeper are event driven. It is difficult for the management server to ensure that it has all the discovery updates for all the clusters given that clusters are also dynamic in nature before it claims it self ready. Let us assume a case which is failing in our deployments - There are two clusters - cluster x and cluster y that are in warming state when an Envoy connected to a management server instance. Envoy sends EDS requests for x and y. When the management server gets update from its discovery backend for x - we will send back the `ClusterLoadAssignment` for cluster x and that response does not have reference to cluster y. Because of the way Envoy works today, on this EDS response both x and y finish warming - x with the hosts sent, y with empty hosts. Later say after 5 mins  as an example our discovery backend sends notification of y hosts and we update it with another EDS response to Envoy. Envoy now updates the hosts correctly. During this period, the requests for cluster y would fail - I think the fundamental problem here is that Envoy is finishing warming of cluster y when management server did not send any EDS response for it - which should be fixed?

Regarding host copy approach, that originated from @htuch 's idea here https://github.com/envoyproxy/envoy/issues/5168#issuecomment-460296843.  Since this copy is just limited to this specific case, it may be OK.

If there is a semantic concern on hosts copy and may lead to edge case issues, we can think about alternatives. Couple of alternatives that come to my mind are 

1. change grpc_mux_impl functionality here https://github.com/envoyproxy/envoy/blob/master/source/common/config/grpc_mux_impl.cc#L171 to call onConfigUpdate only if there are resources. But as I mentioned as mentioned above it had some side effects during Envoy startup and we had to revert it.
2. If the cluster is in warming state, do not finish it's initialization - so basically do not call `onPreInitComplete` here https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/eds.cc#L50 on empty update. So this ensure cluster warming would only be complete when EDS response for that cluster comes. Not sure if there will be any side effects to this. @htuch may know better.

Reply:
So, I think we should clarify whether this is about update warming only or initial warming. I thought it was the former, but in the above comment from @ramaraochavali, this seems like it applies to initial warming as well.

In general, I think the main gap I see in a common understanding here is what does "warming" mean for clusters or listeners that are updating? What are the guarantees we provide when update warming is completed? Do we demand that the management server send an updated RDS/EDS or not? If we can answer this, then other issues become simpler.

Reply:
@htuch you are right. our problem of hosts getting cleared only happens for updating warming case. Initial warming since there does not exist any hosts, we do not see any problem. Sorry if my above comment created that confusion.

Reply:
> It is difficult for the management server to ensure that it has all the discovery updates for all the clusters given that clusters are also dynamic in nature before it claims it self ready.

I see what you are saying, and I agree we have some problems to fix here, but this still doesn't make sense to me. What do you do if a new Envoy comes up and connects to management server that doesn't have a complete knowledge of the topology? How would that work? Doesn't your management server need to initialize with the most recent data and then merge in updates and send them out as they are available?

Of your proposed solutions, my vote is to focus on either (1) or (2), that seems much more correct to me than what we are proposing here. @htuch @snowp thoughts?

Reply:
> What do you do if a new Envoy comes up and connects to management server that doesn't have a complete knowledge of the topology? How would that work? Doesn't your management server need to initialize with the most recent data and then merge in updates and send them out as they are available?

With the readiness check in our management we ensure the management server is upto date with the latest cluster information and latest cluster information is sent to newly connected Envoys. As far as EDS updates go, we send whatever latest information we have in the management server at the point in time when the Envoy connected - rest would go in the async mode as and when they are available. With retries in place we do not see a big problem with this.

> Of your proposed solutions, my vote is to focus on either (1) or (2), that seems much more correct to me than what we are proposing here

If we do not go with what is here in this PR, my vote would be for us to try (2) because as I mentioned earlier (1) has been tried for some other reason has been found to have some regressions. However I think the question that @htuch raised above still need to be answered so that we go in the direction that is aligned with future. 
Also slight addition to his question of "
Do we demand that the management server send an updated RDS/EDS or not?" - "Do we demand that the management server send an updated RDS/EDS or not in a **single** response"  because that is what is causing problem here?

Based on what we decide here I will change the direction of this PR.

Reply:
I agree with @htuch that this boils down to how we want cluster warming to work wrt EDS updates. If we want cluster warming to just rebuild the cluster and continue to use the existing endpoints, then something similar to what we're doing in this PR seems necessary. 

From the beginning I was for the idea of delaying cluster warming until an EDS response came back, but @htuch brought up some good points on the make before break policy in the issue so now I'm less certain about what the right thing to do is.

To me this seems like primarily a management server concern, but I can understand the desire for less strict management servers like what @ramaraochavali is describing. FWIW We're using Zookeeper for service discovery data on our end and we've had no issues with this, presumably due to how we generate both CDS and EDS from the same ZK data. 

If we are going to require an EDS response it seems kinda strange to require it only on cluster warming and not cluster init, so (1) seems like right thing to me, but from the regression that caused it seems like people are expecting to be able to initialize from an empty EDS response, so that'd be a breaking change. We might be stuck with (2).

Reply:
> FWIW We're using Zookeeper for service discovery data on our end and we've had no issues with this, presumably due to how we generate both CDS and EDS from the same ZK data.

That is one and another point to consider here is this happens only when cluster property change (and hence goes in to warming mode) and the control plane is going through rolling upgrades. In other cases we are also fine.

Another thing longer term to consider is what cluster change need warming i.e. EDS updates for example I change some timeout or some other unrelated setting - why do we have to go through warmup cycle (I understand for simplicity we are doing that by just computing the hash on config)? or for cluster update case can we just skip warming and put the onus management server to send EDS updates if the cluster change needs updated end points?

Reply:
> If we want cluster warming to just rebuild the cluster and continue to use the existing endpoints, then something similar to what we're doing in this PR seems necessary

IMO this is not what we want. If a cluster is being warmed it is either new or different (as defined by config) and as such I think it needs to start empty and get hosts via an EDS fetch.

> If we are going to require an EDS response it seems kinda strange to require it only on cluster warming and not cluster init, so (1) seems like right thing to me, but from the regression that caused it seems like people are expecting to be able to initialize from an empty EDS response, so that'd be a breaking change. We might be stuck with (2)

I agree, and IMO we should require a named response before finishing init. I think this might be less painful if https://github.com/envoyproxy/envoy/issues/5862 is implemented (initial fetch timeout to release a cluster if the management server does not respond).

Reply:
>  IMO this is not what we want. If a cluster is being warmed it is either new or different (as defined by config) and as such I think it needs to start empty and get hosts via an EDS fetch.

This all boils down to (im)mutability. It seems hard to guarantee consistency in the presence of mutability of clusters, since, you might in the case of CDS update, still have a stale EDS on the server. So, even if you do wait for an EDS update, you might not get anything meaningful in response. So, if we delay warming and force an EDS update for only a weak improvement in the consistency of mutable CDS clusters.

The management server can still do an EDS update immediately after CDS, even if we reuse existing EDS resources, we just get a period of potential inconsistency.

I think it comes down to whether a management server can guarantee whether an EDS update following a CDS update truly reflect the new CDS state. In some setups this is true, but in others it isn't. 

A more robust pattern would be to force folks to make all resources immutable. I know this is a broader API design question, but it sure would simplify.

Reply:
Maybe one way to simplify this discussion is to come up with a bunch of user stories, e.g. "As an Envoy user, I want to upgrade a service from HTTP to HTTPS.. the recommended steps are ...".

Reply:
> The management server can still do an EDS update immediately after CDS, even if we reuse existing EDS resources, we just get a period of potential inconsistency.

I think it comes down to whether a management server can guarantee whether an EDS update following a CDS update truly reflect the new CDS state. In some setups this is true, but in others it isn't.

I agree with this and it seems potential inconsistency can not be avoided in all cases.

I think we need to think about a short term solution and longer term solutions like immutable resources etc

For short term we have couple of solutions
- Reuse existing EDS resources till management server provides a new set of resources matching the CDS update (Whether this CDS update requires a new set of EDS resources is another question - which we are assuming right now that all CDS updates require EDS response which is kind  of looks not correct)
- Do not finish warming till a named response comes for updated warming case

Do we have any other ideas?

Longer term, for sure we should come up with bunch of user stories and think about solutions like immutable resources etc as @htuch  mentioned.

Reply:
@ramaraochavali take a look at a convo I had with @htuch in Envoy slack: https://envoyproxy.slack.com/archives/CEFDKQ3RQ/p1551391835017700

I'm pretty strongly of the opinion that we should not finish init/warming until we explicitly receive a named response for an EDS fetch. I think this is the most clear solution and IMO the behavior that most people would expect. 

@htuch brings up the valid point that this may not work in all cases if the hosts have changed, but IMO this is very rare, and we should offer guidance in the XDS docs that this edge case should be handled by a cluster rename and correct sequencing.

Reply:
@mattklein123 @htuch @snowp  Created this doc https://docs.google.com/document/d/1Ca4YX9XNnV9rjTR7U0_xyCxUDKfoSun8pWSUNO9N-tY/edit?usp=sharing so that we can discuss there.

Reply:
OK. So since we got consensus on finishing warming only on named responses, I am going to close this PR and open another PR with a fix.